avocado/plugins/diff.py

```diff
@@ -0,0 +1,387 @@
+
+    """
+    Implements the avocado 'diff' subcommand
+    """
+
+    name = 'diff'
+    description = 'Shows the difference between 2 jobs.'
+
+    def configure(self, parser):
+        """
+        Add the subparser for the diff action.
+
+        :param parser: Main test runner parser.
+        """
+        parser = super(Diff, self).configure(parser)
+
+        parser.add_argument("jobids",
+                            default=[], nargs=2,
+                            metavar="<JOB>",
+                            help='Provide two Jobs, identified by their Job '
+                            '(partial) ID or test results directory.')
+
+        parser.add_argument('--html', type=str,
+                            dest='html_output',
+                            metavar='FILE',
+                            help='Enable HTML output to the FILE where the '
+                            'result should be written. The value - (output to '
+                            'stdout) is not supported since not all HTML '
+                            'resources can be embedded into a single file '
+                            '(page resources will be copied to the output '
+                            'file dir)')
+
+        parser.add_argument('--open-browser',
+                            action='store_true',
+                            default=False,
+                            help='Generate and open a HTML report in your '
+                            'preferred browser. If no --html file is '
+                            'provided, we create a temporary file. '
+                            'Current: False.')
+
+        parser.add_argument('--diff-filter',
+                            dest='diff_filter',
+                            type=self._validate_filters,
+                            default=['all'],
+                            help='Diff filter: all,(-)cmdline,(-)time,'
+                            '(-)variants,(-)results,(-)config,(-)sysinfo '
+                            '(defaults to all).')
+
+        parser.add_argument('--paginator',
+                            choices=('on', 'off'), default='on',
+                            help='Turn the paginator on/off. '
+                            'Current: %(default)s')
+
+        self.term = output.TERM_SUPPORT
+
+    def run(self, args):
+        job1_dir, job1_id = self._setup_job(args.jobids[0])
+        job2_dir, job2_id = self._setup_job(args.jobids[1])
+
+        job1_data = self._get_job_data(job1_dir)
+        job2_data = self._get_job_data(job2_dir)
+
+        job1_results = []
+        job2_results = []
+
+        if (('all' in args.diff_filter or 'cmdline' in args.diff_filter) and
+           '-cmdline' not in args.diff_filter):
+
+            cmdline1 = self._get_command_line(job1_dir)
+            cmdline2 = self._get_command_line(job2_dir)
+
+            if self._has_diff([cmdline1], [cmdline2]):
+                command_line_header = ['\n', 'COMMAND LINE\n']
```

Comment:
On Wed, Aug 10, 2016 at 05:00:51AM -0700, Lukáš Doktor wrote:
> > +        job2_dir, job2_id = self._setup_job(args.jobids[1])
> > +
> > +        job1_data = self._get_job_data(job1_dir)
> > +        job2_data = self._get_job_data(job2_dir)
> > +
> > +        job1_results = []
> > +        job2_results = []
> > +
> > +        if (('all' in args.diff_filter or 'cmdline' in args.diff_filter) and
> > +           '-cmdline' not in args.diff_filter):
> > +
> > +            cmdline1 = self._get_command_line(job1_dir)
> > +            cmdline2 = self._get_command_line(job2_dir)
> > +
> > +            if self._has_diff([cmdline1], [cmdline2]):
> > +                command_line_header = ['\n', 'COMMAND LINE\n']
> 
> The output is nice and colorful, only the headings are quite
> invisible. I'd consider using bold bright color along with some
> formatting like:
> 
> ```
>  /============\
> | COMMAND LINE |
>  \============/
> ```

We need to be consistent. The avocado "UI" is quite simplistic,
without any bells and whistles. I like the current approach
followed by Amador here, where we give emphasis to the headings
using all-caps.  Lines in a terminal are a precious resource,
let's not waste it with fancy lines around them.

And the output format is a diff and the headings are context.
There's no "bold bright color" in a diff context.

(remember this feature is targeted at developers, all who are
quite used to textual interfaces and the diff format)

With that said, I guess we can all make suggestions, but the
output format is Amador's call.

-- 
Ademar Reis
Red Hat

^[:wq!